[Shipping labels] invalid package dimensions error

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-889

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This adds an error message to the custom package screen for dimension fields, displayed when the entered value is not positive.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
1. Go to Orders.
2. Select an order that is eligible for creating shipping labels.
3. Tap Create shipping label button.
4. Tap Select a Package button.
5. Enter invalid values in the dimension fields.
6. Tap Add Package button.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Steps above.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="400" src="https://github.com/user-attachments/assets/8a3315ba-174f-4806-a02d-4ce61a4167f9" />

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Add invalid dimension string for package creation

irfano committed:
Add invalid dimension validation to custom packages

The `invalidDimensions` property checks if any of the length, width, or height are less than or equal to zero.

irfano committed:
Show error message when custom package dimensions are invalid

irfano committed:
Remove unused field from PackageCreationResponse

irfano committed:
Fix WooShippingLabelPackageCreationViewModelTest

irfano committed:
Add test for invalid custom package dimensions

irfano committed:
Update release notes

Add a release note for adding an error message to the custom package creation screen for invalid dimensions.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/packages/networking/PackageResponseDTOs.kt:
```diff
@@ -4,7 +4,6 @@ import com.google.gson.annotations.SerializedName
 
 class PackageCreationResponse {
     val custom: List<CustomPackageDTO>? = null
-    val predefined: Map<String, List<String>>? = null
```
This property was unused in the project, so I removed it.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/packages/ui/UIModels.kt:
```diff
@@ -107,6 +108,11 @@ data class CustomPackageCreationData(
             return name.isNotNullOrEmpty() && weight.isNotNullOrEmpty()
         }
 
+    val invalidDimensions: Boolean
+        get() = (length.toDoubleOrNull() ?: 0.0) <= 0.0 ||
+            (width.toDoubleOrNull() ?: 0.0) <= 0.0 ||
+            (height.toDoubleOrNull() ?: 0.0) <= 0.0
```
With a physical keyboard, users can enter letters in the dimension fields. The `toDoubleOrNull()` check handles this as well: it shows an error if the fields contain anything other than numbers.

irfano commented on WooCommerce/src/main/res/values/strings.xml:
```diff
@@ -4641,6 +4641,7 @@
     <string name="woo_shipping_labels_package_creation_height">Height</string>
     <string name="woo_shipping_labels_package_creation_weight">Weight</string>
     <string name="woo_shipping_labels_package_creation_package_name">Package Name</string>
+    <string name="woo_shipping_labels_package_creation_invalid_dimension">Package dimensions should all be larger than 0.</string>
```
I copied this message from iOS.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/packages/WooShippingLabelPackageCreationViewModel.kt:
```diff
@@ -175,16 +175,21 @@ class WooShippingLabelPackageCreationViewModel @Inject constructor(
-                    )
+        when {
+            customPackage.invalidDimensions -> _viewState.value = _viewState.value.copy(
+                customPackageCreationData = _viewState.value.customPackageCreationData.copy(
+                    invalidDimensionError = true
                 )
```
I think we can improve the UX a little bit. Right now, we don't clear the error even when the user enters valid   dimensions:

https://github.com/user-attachments/assets/c9d58d40-52d4-4a62-9044-a58084582b4d

Maybe we can clear the error when the user edits the dimensions?

irfano replied:
Nice suggestion. I applied it: 232be2b4f84ecbe3d5ad5bd969d0666f16aa10fb

hichamboushaba commented on RELEASE-NOTES.txt:
```diff
@@ -19,6 +19,7 @@
 - [*] Shipping Labels: Improved handling of address validation errors [https://github.com/woocommerce/woocommerce-android/pull/14468]
 - [*] Shipping Labels: Improved the error message for empty shipping rates list [https://github.com/woocommerce/woocommerce-android/pull/14433]
 - [*] Shipping Labels: Update button logic on the Hazardous Materials screen [https://github.com/woocommerce/woocommerce-android/pull/14488]
+- [*] Shipping Labels: Add error message for invalid dimensions errors to custom package creation screen [https://github.com/woocommerce/woocommerce-android/pull/14499]
```
Please make sure to move this note to the version 23.2

irfano replied:
I’ve moved it. Thanks for the reminder. 321456b8c23aad8080a2caf83bd14f7e495fdf7e

hichamboushaba reviewed (approved):
Nice work @irfano, I left some comments, but I'm pre-approving.

I also wanted to report this minor, unrelated issue: the text color of the "Add package" button is not correct in dark mode, it should be white. Let's use `WCColoredButton` to avoid these inconsistencies. Can you update it as part of this PR, even if it's unrelated?

<img width="360" alt="Image" src="https://github.com/user-attachments/assets/a33dea3b-b188-4b1f-aa50-ad156724d02a" />

irfano committed:
Fix button color on shipping package screens

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-889-invalid-package-dimensions-error

irfano committed:
Revalidate dimension errors on dimension change

When a dimension (length, width, or height) of a custom package changes, and there was a pre-existing invalid dimension error, the error state is re-evaluated.
This ensures that if the change resolves the error, the error state is cleared.

irfano committed:
Fix detekt error

irfano committed:
Update RELEASE-NOTES.txt for 23.2 release

The release notes are updated to include the change for adding an error message for invalid dimensions in custom package creation for shipping labels.

irfano commented:
Thank you for your feedback! I’ve applied all of them.

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-889-invalid-package-dimensions-error
